File: foundations-jdbc-scala/src/scala/dev/typr/foundationssc/OperationRead.scala
package dev.typr.foundationssc

import java.time.Duration

sealed trait OperationRead[Out] extends Operation[Out]:

  override def underlying: dev.typr.foundations.OperationRead[Out]

  def run(using conn: dev.typr.foundations.ConnectionRead): Out =
    conn.execute(underlying)

  def transactRead(transactor: Transactor): Out =
    transactor.execute(this)

  override def map[B](f: Out => B): OperationRead[B] =
    OperationRead.Mapped(this, f)

  override def voided(): OperationRead[Unit] = map(_ => ())

  def combine[B](other: OperationRead[B]): OperationRead[(Out, B)] =
    OperationRead.CombineRead(this, other)

  def combineWith[B, R](other: OperationRead[B])(f: (Out, B) => R): OperationRead[R] =
    combine(other).map { case (a, b) => f(a, b) }

  def combineWith[B, C, R](b: OperationRead[B], c: OperationRead[C])(f: (Out, B, C) => R): OperationRead[R] =
    combine(b).combine(c).map { case ((a, b), c) => f(a, b, c) }

  def combineWith[B, C, D, R](b: OperationRead[B], c: OperationRead[C], d: OperationRead[D])(f: (Out, B, C, D) => R): OperationRead[R] =
    combine(b).combine(c).combine(d).map { case (((a, b), c), d) => f(a, b, c, d) }

  def combineWith[B, C, D, E, R](b: OperationRead[B], c: OperationRead[C], d: OperationRead[D], e: OperationRead[E])(
      f: (Out, B, C, D, E) => R
  ): OperationRead[R] =
    combine(b).combine(c).combine(d).combine(e).map { case ((((a, b), c), d), e) => f(a, b, c, d, e) }

  def combineWith[B, C, D, E, F, R](b: OperationRead[B], c: OperationRead[C], d: OperationRead[D], e: OperationRead[E], f: OperationRead[F])(
      fn: (Out, B, C, D, E, F) => R
  ): OperationRead[R] =
    combine(b).combine(c).combine(d).combine(e).combine(f).map { case (((((a, b), c), d), e), f) => fn(a, b, c, d, e, f) }

  def productL[B](other: OperationRead[B]): OperationRead[Out] =
    combine(other).map(_._1)

  def andThen[B](next: Out => OperationRead[B]): OperationRead[B] =
    OperationRead.ThenRead(this, next)

  override def named(name: String): OperationRead[Out] =
    OperationRead.JavaWrapped(underlying.named(name))

  override def timeout(timeout: Duration): OperationRead[Out] =
    OperationRead.JavaWrapped(underlying.timeout(timeout))

  override def withListener(listener: QueryListener): OperationRead[Out] =
    OperationRead.JavaWrapped(underlying.withListener(listener))

object OperationRead:

  class Query[Out](override val underlying: dev.typr.foundations.OperationRead.Query[Out]) extends OperationRead[Out]

  object Query:
    def apply[Out](query: Fragment, parser: ResultSetParser[Out]): Query[Out] =
      new Query(new dev.typr.foundations.OperationRead.Query(query.underlying, parser.underlying))

  class Streaming[Row](private val java: dev.typr.foundations.OperationRead.Streaming[Row]) extends OperationRead[Cursor[Row]]:
    override val underlying: dev.typr.foundations.OperationRead[Cursor[Row]] =
      java.map[Cursor[Row]]((jc: dev.typr.foundations.Cursor[Row]) => new Cursor(jc))

  object Streaming:
    def apply[Row](query: Fragment, codec: RowCodec[Row], fetchSize: Int): Streaming[Row] =
      new Streaming(new dev.typr.foundations.OperationRead.Streaming(query.underlying, codec.underlying, fetchSize))

  class Pure[T](override val underlying: dev.typr.foundations.OperationRead.Pure[T]) extends OperationRead[T]

  private[foundationssc] class Mapped[A, B](source: OperationRead[A], f: A => B) extends OperationRead[B]:
    override val underlying: dev.typr.foundations.OperationRead[B] =
      source.underlying.map[B]((a: A) => f(a))

  private[foundationssc] class CombineRead[A, B](first: OperationRead[A], second: OperationRead[B]) extends OperationRead[(A, B)]:
    override val underlying: dev.typr.foundations.OperationRead[(A, B)] =
      first.underlying
        .combine(second.underlying)
        .map[(A, B)]((t: dev.typr.foundations.Tuple.Tuple2[A, B]) => (t._1(), t._2()))

  private[foundationssc] class ThenRead[A, B](source: OperationRead[A], continuation: A => OperationRead[B]) extends OperationRead[B]:
    override val underlying: dev.typr.foundations.OperationRead[B] =
      source.underlying.thenRead((a: A) => continuation(a).underlying)

  private[foundationssc] class IfEmptyRead[T](check: OperationRead[Option[T]], fallback: OperationRead[T]) extends OperationRead[T]:
    override val underlying: dev.typr.foundations.OperationRead[T] =
      dev.typr.foundations.OperationRead.ifEmpty[T](
        check.underlying.map(opt => if opt.isDefined then java.util.Optional.of(opt.get) else java.util.Optional.empty[T]()),
        fallback.underlying
      )

  private[foundationssc] class ConfiguredRead[Out](override val underlying: dev.typr.foundations.OperationRead[Out]) extends OperationRead[Out]

  private[foundationssc] class JavaWrapped[Out](override val underlying: dev.typr.foundations.OperationRead[Out]) extends OperationRead[Out]

  def pure[T](value: T): OperationRead[T] = Pure(new dev.typr.foundations.OperationRead.Pure(value))

  def sequence[T](operations: List[OperationRead[T]]): OperationRead[List[T]] =
    if operations.isEmpty then return pure(Nil)
    var result: OperationRead[List[T]] = operations.head.map(t => List(t))
    for op <- operations.tail do result = result.combine(op).map(t => t._1 :+ t._2)
    result

  def allOf(operations: OperationRead[?]*): OperationRead[Unit] =
    if operations.isEmpty then return pure(())
    var result: OperationRead[Unit] = operations.head.voided()
    for op <- operations.tail do result = result.productL(op)
    result

  def ifEmpty[T](check: OperationRead[Option[T]], fallback: OperationRead[T]): OperationRead[T] =
    IfEmptyRead(check, fallback)

File: foundations-jdbc/src/java/dev/typr/foundations/OperationRead.java
package dev.typr.foundations;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.function.BiFunction;

public sealed interface OperationRead<Out> extends Operation<Out>
    permits OperationRead.Query,
        OperationRead.Streaming,
        OperationRead.Pure,
        OperationRead.Mapped,
        OperationRead.Combine,
        OperationRead.IfEmpty,
        OperationRead.Then,
        OperationRead.Configured {

  /** Run this read operation on a read-only connection. */
  default Out run(ConnectionRead conn) {
    return conn.execute(this);
  }

  /**
   * Re-declares the inherited {@link Operation#run(Connection)} on this trait so Scala 3's
   * overload resolver sees both {@code run(Connection)} and {@code run(ConnectionRead)} declared
   * at the same level — then specificity ({@code Connection <: ConnectionRead}) picks the more
   * specific overload for {@code c: Connection} call sites without ambiguity. Java and Kotlin
   * always picked the most-specific overload regardless; this redeclaration is purely for the
   * Scala 3 overload resolver. Pure delegation; same body as the inherited default.
   */
  @Override
  default Out run(Connection conn) {
    return conn.execute(this);
  }

  /**
   * Execute this operation using the given transactor.
   *
   * @param transactor the transactor to use
   * @return the operation result
   * @throws DatabaseException if a database error occurs
   */
  default Out transactRead(Transactor transactor) {
    return transactor.execute(this);
  }

  @Override
  default <B> OperationRead<B> map(SqlFunction<Out, B> f) {
    return new Mapped<>(this, f);
  }

  default <B> OperationRead<Tuple.Tuple2<Out, B>> combine(OperationRead<B> other) {
    return new Combine<>(this, other);
  }

  default <B, C> OperationRead<Tuple.Tuple3<Out, B, C>> combine(
      OperationRead<B> b, OperationRead<C> c) {
    return combine(b)
        .combine(c)
        .map(t -> new Tuple.Tuple3.Impl<>(t._1()._1(), t._1()._2(), t._2()));
  }

  default <B, C, D> OperationRead<Tuple.Tuple4<Out, B, C, D>> combine(
      OperationRead<B> b, OperationRead<C> c, OperationRead<D> d) {
    return combine(b, c)
        .combine(d)
        .map(t -> new Tuple.Tuple4.Impl<>(t._1()._1(), t._1()._2(), t._1()._3(), t._2()));
  }

  default <B, C, D, E> OperationRead<Tuple.Tuple5<Out, B, C, D, E>> combine(
      OperationRead<B> b, OperationRead<C> c, OperationRead<D> d, OperationRead<E> e) {
    return combine(b, c, d)
        .combine(e)
        .map(
            t ->
                new Tuple.Tuple5.Impl<>(
                    t._1()._1(), t._1()._2(), t._1()._3(), t._1()._4(), t._2()));
  }

  default <B, C, D, E, F> OperationRead<Tuple.Tuple6<Out, B, C, D, E, F>> combine(
      OperationRead<B> b,
      OperationRead<C> c,
      OperationRead<D> d,
      OperationRead<E> e,
      OperationRead<F> f) {
    return combine(b, c, d, e)
        .combine(f)
        .map(
            t ->
                new Tuple.Tuple6.Impl<>(
                    t._1()._1(), t._1()._2(), t._1()._3(), t._1()._4(), t._1()._5(), t._2()));
  }

  default <B, R> OperationRead<R> combineWith(
      OperationRead<B> other, BiFunction<Out, B, R> combine) {
    return combine(other).map(t -> combine.apply(t._1(), t._2()));
  }

  default <B, C, R> OperationRead<R> combineWith(
      OperationRead<B> b, OperationRead<C> c, Functions.Function3<Out, B, C, R> combine) {
    return combine(b, c).map(t -> combine.apply(t._1(), t._2(), t._3()));
  }

  default <B, C, D, R> OperationRead<R> combineWith(
      OperationRead<B> b,
      OperationRead<C> c,
      OperationRead<D> d,
      Functions.Function4<Out, B, C, D, R> combine) {
    return combine(b, c, d).map(t -> combine.apply(t._1(), t._2(), t._3(), t._4()));
  }

  default <B, C, D, E, R> OperationRead<R> combineWith(
      OperationRead<B> b,
      OperationRead<C> c,
      OperationRead<D> d,
      OperationRead<E> e,
      Functions.Function5<Out, B, C, D, E, R> combine) {
    return combine(b, c, d, e).map(t -> combine.apply(t._1(), t._2(), t._3(), t._4(), t._5()));
  }

  default <B, C, D, E, F, R> OperationRead<R> combineWith(
      OperationRead<B> b,
      OperationRead<C> c,
      OperationRead<D> d,
      OperationRead<E> e,
      OperationRead<F> f,
      Functions.Function6<Out, B, C, D, E, F, R> combine) {
    return combine(b, c, d, e, f)
        .map(t -> combine.apply(t._1(), t._2(), t._3(), t._4(), t._5(), t._6()));
  }

  default <B> OperationRead<Out> productL(OperationRead<B> other) {
    return combine(other).map(t -> t._1());
  }

  default <B> OperationRead<B> thenRead(java.util.function.Function<Out, OperationRead<B>> next) {
    return new Then<>(this, next);
  }

  @Override
  default OperationRead<Void> voided() {
    return map(ignored -> null);
  }

  @Override
  default OperationRead<Out> named(String name) {
    if (this instanceof Configured<Out> c) {
      return new Configured<>(c.inner(), Optional.of(name), c.timeout(), c.listener());
    }
    return new Configured<>(this, Optional.of(name), Optional.empty(), Optional.empty());
  }

  @Override
  default OperationRead<Out> timeout(Duration timeout) {
    if (this instanceof Configured<Out> c) {
      return new Configured<>(c.inner(), c.name(), Optional.of(timeout), c.listener());
    }
    return new Configured<>(this, Optional.empty(), Optional.of(timeout), Optional.empty());
  }

  @Override
  default OperationRead<Out> withListener(QueryListener listener) {
    if (this instanceof Configured<Out> c) {
      return new Configured<>(c.inner(), c.name(), c.timeout(), Optional.of(listener));
    }
    return new Configured<>(this, Optional.empty(), Optional.empty(), Optional.of(listener));
  }

  static <T> OperationRead<T> pure(T value) {
    return new Pure<>(value);
  }

  static <T> OperationRead<List<T>> sequence(List<OperationRead<T>> ops) {
    if (ops.isEmpty()) return pure(List.of());
    OperationRead<List<T>> result = ops.getFirst().map(List::of);
    for (int i = 1; i < ops.size(); i++) {
      result =
          result
              .combine(ops.get(i))
              .map(
                  t -> {
                    var list = new ArrayList<>(t._1());
                    list.add(t._2());
                    return Collections.unmodifiableList(list);
                  });
    }
    return result;
  }

  static OperationRead<Void> allOf(OperationRead<?>... ops) {
    if (ops.length == 0) return pure(null);
    OperationRead<Void> result = ops[0].voided();
    for (int i = 1; i < ops.length; i++) {
      result = result.productL(ops[i]);
    }
    return result;
  }

  static <T> OperationRead<T> ifEmpty(OperationRead<Optional<T>> check, OperationRead<T> fallback) {
    return new IfEmpty<>(check, fallback);
  }

  record Query<Out>(Fragment query, ResultSetParser<Out> parser) implements OperationRead<Out> {
    @Override
    public String description(boolean verbose) {
      return "Query: " + query.renderInterpolated();
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Streaming<Row>(Fragment query, RowCodec<Row> codec, int fetchSize)
      implements OperationRead<Cursor<Row>> {
    @Override
    public String description(boolean verbose) {
      return "Streaming: " + query.renderInterpolated();
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Mapped<A, B>(OperationRead<A> source, SqlFunction<A, B> f) implements OperationRead<B> {
    @Override
    public String description(boolean verbose) {
      return "Mapped(" + source.description(verbose) + ")";
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Pure<T>(T value) implements OperationRead<T> {
    @Override
    public String description(boolean verbose) {
      return "Pure(" + value + ")";
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Combine<A, B>(OperationRead<A> first, OperationRead<B> second)
      implements OperationRead<Tuple.Tuple2<A, B>> {
    @Override
    public String description(boolean verbose) {
      return "Combine(" + first.description(verbose) + ", " + second.description(verbose) + ")";
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record IfEmpty<T>(OperationRead<Optional<T>> check, OperationRead<T> fallback)
      implements OperationRead<T> {
    @Override
    public String description(boolean verbose) {
      return "IfEmpty(" + check.description(verbose) + ", " + fallback.description(verbose) + ")";
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Then<A, B>(
      OperationRead<A> source, java.util.function.Function<A, OperationRead<B>> continuation)
      implements OperationRead<B> {
    @Override
    public String description(boolean verbose) {
      return "Then(" + source.description(verbose) + " -> ?)";
    }

    @Override
    public String toString() {
      return description(false);
    }
  }

  record Configured<Out>(
      OperationRead<Out> inner,
      Optional<String> name,
      Optional<Duration> timeout,
      Optional<QueryListener> listener)
      implements OperationRead<Out> {
    @Override
    public String description(boolean verbose) {
      if (name.isPresent()) {
        return verbose ? name.get() + "\n" + inner.description(false) : name.get();
      }
      return inner.description(verbose);
    }

    @Override
    public String toString() {
      return description(false);
    }
  }
}
